sgkit_bgen/bgen_reader.py

```diff
@@ -0,0 +1,186 @@
+
+            df = mf.create_variants()
+            if persist:
+                df = df.persist()
+            variant_arrs = _to_dict(df, dtype=VARIANT_ARRAY_DTYPE)
+
+            self.variant_id = variant_arrs["id"]
+            self.contig = variant_arrs["chrom"]
+            self.pos = variant_arrs["pos"]
+
+            def split_alleles(alleles, block_info=None):
+                if block_info is None or len(block_info) == 0:
+                    return alleles
+
+                def split(allele_row):
+                    return np.array(allele_row[0].split(",")[:2], dtype="S1")
+
+                return np.apply_along_axis(split, 1, alleles[:, np.newaxis])
+
+            self.variant_alleles = variant_arrs["allele_ids"].map_blocks(split_alleles)
+
+        with bgen_file(self.path) as bgen:
+            sample_path = self.path.with_suffix(".sample")
+            if sample_path.exists():
+                self.sample_id = read_samples_file(sample_path, verbose=False)
+            else:
+                if bgen.contain_samples:
+                    self.sample_id = bgen.read_samples()
+                else:
+                    self.sample_id = generate_samples(bgen.nsamples)
+
+        self.shape = (self.n_variants, len(self.sample_id))
+        self.dtype = dtype
+        self.ndim = 2
+
+    def __getitem__(self, idx):
+        if not isinstance(idx, tuple):
+            raise IndexError(  # pragma: no cover
+                f"Indexer must be tuple (received {type(idx)})"
+            )
+        if len(idx) != self.ndim:
+            raise IndexError(  # pragma: no cover
+                f"Indexer must be two-item tuple (received {len(idx)} slices)"
+            )
+
+        if idx[0].start == idx[0].stop:
+            return np.empty((0, 0), dtype=self.dtype)
+
+        start_partition = idx[0].start // self.partition_size
+        start_partition_offset = idx[0].start % self.partition_size
+        end_partition = (idx[0].stop - 1) // self.partition_size
+        end_partition_offset = (idx[0].stop - 1) % self.partition_size
+
+        all_vaddr = []
+        with bgen_metafile(self.metafile_filepath) as mf:
+            for i in range(start_partition, end_partition + 1):
+                partition = mf.read_partition(i)
+                start_offset = start_partition_offset if i == start_partition else 0
+                end_offset = (
+                    end_partition_offset + 1
+                    if i == end_partition
+                    else self.partition_size
+                )
+                vaddr = partition["vaddr"].tolist()
+                all_vaddr.extend(vaddr[start_offset:end_offset])
+
+        with bgen_file(self.path) as bgen:
+            genotypes = [bgen.read_genotype(vaddr) for vaddr in all_vaddr]
+            all_probs = [genotype["probs"] for genotype in genotypes]
+            d = [(2 * probs[:, -1] + probs[:, 1]) for probs in all_probs]
+            return np.stack(d)[:, idx[1]]
```

Comment:
I don't suppose bgen_reader has a way to limit the genotype data fetched for a subset of samples right?

My mental model for this function is that dask may send several column slices (for subsets of samples) for the same row slice and this reader will load *all* the samples into memory for the row slice and then slice off the columns here at the end.  

It looks like the bytes for all samples for a specific variant should be sequential on disk in the bgen spec so I'm wondering if there's a way to use the column slice to tell bgen reader to seek to that subset instead of loading all of them each time.  I can see that being an important optimization.

Reply:
I agree, however it doesn't seem to be possible with the current implementation of bgen_reader. I've opened https://github.com/pystatgen/sgkit-bgen/issues/2 so we can investigate further.